fix(ivy): introduce host-specific styling instructions

matsko opened the pull request:
This patch is the first of a few patches which separates the
styling logic between template bindings (e.g. `<div [style]`)
from host bindings (e.g. `@HostBinding('style')`). This patch
in particular introduces a series of host-specific styling
instructions and changes the existing set of template styling
instructions not to accept directives. The underyling code (which
communicates with the styling algorithm) still works as it did
before.

This PR also separates the styling instruction code into a separate
file and moves over all other instructions into an dedicated
instructions directory.

matsko commented:
http://test/OCL:238467080:BASE:238467346:1552583824289:ab8a0ca2

mhevery commented on packages/common/src/directives/ng_class.ts:
```diff
@@ -35,11 +35,11 @@ export const ngClassDirectiveDef__POST_R3__ = ɵdefineDirective({
   hostBindings: function(rf: ɵRenderFlags, ctx: any, elIndex: number) {
     if (rf & ɵRenderFlags.Create) {
-      ɵelementStyling(null, null, null, ctx);
+      ɵelementHostStyling(ctx);
```
There is no need to pass in `ctx` since it can be looked by the instruction using `geLView()[CONTEXT]`

mhevery commented on packages/common/src/directives/ng_class.ts:
```diff
@@ -35,11 +35,11 @@ export const ngClassDirectiveDef__POST_R3__ = ɵdefineDirective({
     if (rf & ɵRenderFlags.Update) {
-      ɵelementStylingMap(elIndex, ctx.getValue(), null, ctx);
-      ɵelementStylingApply(elIndex, ctx);
+      ɵelementHostStylingMap(ctx, elIndex, ctx.getValue());
```
There is no need to pass in `ctx` since it can be looked by the instruction using `geLView()[CONTEXT]`

There is no need to pass in `elIndex` because it can be looked up by (i think) `getPreviousOrParentTNode().index`

mhevery commented on tools/public_api_guard/global_utils.d.ts:
```diff
@@ -15,5 +15,3 @@ export declare function getPlayers(ref: ComponentInstance | DirectiveInstance |
 export declare function getRootComponents(target: {}): any[];
 
 export declare function getViewComponent<T = {}>(element: Element | {}): T | null;
-
```
This should be reverted. Not sure why you are removing it, seems like an accidental removal.

mhevery commented on packages/core/src/render3/instructions/styling_instructions.ts:
```diff
@@ -0,0 +1,427 @@
+
+
+/**
+ * Allocates style and class binding properites on the element during creation mode.
```
```suggestion
 * Allocates style and class binding properties on the element during creation mode.
```

mhevery commented on packages/core/src/render3/instructions/styling_instructions.ts:
```diff
@@ -0,0 +1,427 @@
+ */
+export function elementStyling(
+    classBindingNames?: string[] | null, styleBindingNames?: string[] | null,
+    styleSanitizer?: StyleSanitizeFn | null, directive?: {}): void {
```
remove `directive` argument

mhevery commented on packages/core/src/render3/instructions/styling_instructions.ts:
```diff
@@ -0,0 +1,427 @@
+}
+
+/**
+ * Allocates style and class binding properites on the host element during creation mode
```
```suggestion
 * Allocates style and class binding properties on the host element during creation mode
```

mhevery commented on packages/core/src/render3/instructions/styling_instructions.ts:
```diff
@@ -0,0 +1,427 @@
+    tNode.stylingTemplate = createEmptyStylingContext();
+  }
+
+  // this will make sure that the root directive (the template) will always be
```
awkward, can you reword

mhevery commented on packages/core/src/render3/instructions/styling_instructions.ts:
```diff
@@ -0,0 +1,427 @@
+ * component. Note that this is only used for binding values (see `elementHostAttrs`
+ * to learn how to assign static styling values to the host element).
+ *
+ * @param directive The directive or component instance that the styling is associated with.
```
remove

mhevery commented on packages/core/src/render3/instructions/styling_instructions.ts:
```diff
@@ -0,0 +1,427 @@
+/**
+ * Update style and/or class bindings using object literals on an element.
+ *
+ * This instruction is meant apply styling via the `[style]="exp"` and `[class]="exp"` template
```
`meant to apply`?

mhevery commented on packages/core/src/render3/instructions/styling_instructions.ts:
```diff
@@ -0,0 +1,427 @@
+ * Update style and/or class host bindings using object literals on an element within the host
+ * bindings function for a directive/component.
+ *
+ * This instruction is meant apply styling via the `@HostBinding('style')` and
```
`meant to apply`?

matsko committed:
refactor(ivy): move instructions into separate directory

matsko committed:
fix(ivy): introduce host-specific styling instructions

This patch is the first of a few patches which separates the
styling logic between template bindings (e.g. <div [style])
from host bindings (e.g. @HostBinding('style')). This patch
in particular introduces a series of host-specific styling
instructions and changes the existing set of template styling
instructions not to accept directives. The underyling code (which
communicates with the styling algorithm) still works as it did
before.

This PR also separates the styling instruction code into a separate
file and moves over all other instructions into an dedicated
instructions directory.

matsko commented:
https://test.corp.google.com/ui#id=OCL:238467080:BASE:238709470:1552684825362:43da7f46

mhevery commented on packages/common/src/directives/ng_class.ts:
```diff
@@ -35,11 +35,11 @@ export const ngClassDirectiveDef__POST_R3__ = ɵdefineDirective({
-      ɵelementStylingMap(elIndex, ctx.getValue(), null, ctx);
-      ɵelementStylingApply(elIndex, ctx);
+      ɵelementHostStylingMap(ctx.getValue());
+      ɵelementHostStylingApply();
```
So much better!

matsko commented:
http://test/OCL:238467080:BASE:239126208:1552972122157:43ad0d9f

matsko commented:
Already passing = https://tap.corp.google.com/ui#id=OCL:238467080:BASE:239126208:1552977675621:86778a56
